source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+public:
+  ScaledRangeTimer(TimerCb callback, ScaledRangeTimerManager& manager)
+      : manager_(manager), callback_(callback),
+        pending_timer_(manager.dispatcher_.createTimer([this] { onPendingTimerComplete(); })) {}
+
+  ~ScaledRangeTimer() override { disableTimer(); }
+
+  void disableTimer() override {
+    struct Dispatch {
+      Dispatch(ScaledRangeTimer& timer) : timer(timer) {}
+      ScaledRangeTimer& timer;
+      void operator()(const Inactive&) {}
+      void operator()(const Pending&) { timer.pending_timer_->disableTimer(); }
+      void operator()(const Active& active) {
+        timer.manager_.disableActiveTimer(active.duration, active.bucket_position);
+      }
+    };
+    absl::visit(Dispatch(*this), state_);
+    state_.emplace<Inactive>();
+  }
+
+  void enableTimer(const std::chrono::milliseconds& min_ms, const std::chrono::milliseconds& max_ms,
+                   const ScopeTrackedObject* scope) override {
+    disableTimer();
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
+    const MonotonicTime latest_trigger = now + max_ms;
+    scope_ = scope;
+    ENVOY_LOG_MISC(info, "enableTimer called on {} for ({}ms, {}ms)", static_cast<void*>(this),
+                   min_ms.count(), max_ms.count());
+    if (min_ms <= std::chrono::milliseconds::zero()) {
+      auto it = manager_.add(*this, max_ms);
+      state_.emplace<Active>(max_ms, it);
+    } else {
+      state_.emplace<Pending>(latest_trigger);
+      pending_timer_->enableTimer(min_ms);
+    }
+  }
+
+  bool enabled() override { return !absl::holds_alternative<Inactive>(state_); }
+
+  void trigger() {
+    ASSERT(absl::holds_alternative<Active>(state_));
+    state_.emplace<Inactive>();
+    if (scope_ == nullptr) {
+      callback_();
+    } else {
+      ScopeTrackerScopeState scope(scope_, manager_.dispatcher_);
+      callback_();
+    }
+    scope_ = nullptr;
+  }
+
+private:
+  friend class ScaledRangeTimerManager;
+
+  struct Inactive {};
+
+  struct Pending {
+    Pending(MonotonicTime latest_trigger) : latest_trigger(latest_trigger) {}
+
+    const MonotonicTime latest_trigger;
+  };
+
+  struct Active {
+    Active(MonotonicTime::duration duration,
+           ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position)
+        : duration(duration), bucket_position(std::move(bucket_position)) {}
+    const MonotonicTime::duration duration;
+    const ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position;
+  };
+
+  void onPendingTimerComplete() {
+    ENVOY_LOG_MISC(info, "pending complete for {}", static_cast<void*>(this));
+    ASSERT(absl::holds_alternative<Pending>(state_));
+    Pending& pending = absl::get<Pending>(state_);
+
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
+    auto it = manager_.add(*this, pending.latest_trigger - now);
+    state_.emplace<Active>(pending.latest_trigger - now, it);
```

Comment:
Should you prefer use of max-min instead of latest_trigger - now ?

The argument is related to reducing the possible values that the range timer manager has to interact with.

Reply:
Reducing the number of possible values doesn't get us anything with the current implementation. Using the latest trigger time makes sure that even if the pending timer fires late, the range timer doesn't overshoot.

Reply:
Getting time from a real clock is almost never the right thing to do in the event loop for performance and consistency reasons.  Sounds like this is something I need to fix since I'm one of the people who is most invested in the behavior of the event loop.

Reply:
My recommendation is that you prefer use of max_configured-min_configured which seems more deterministic by avoiding computations based on approximateMonotonicTime.  If epoll durations and delay are high, we have bigger problems.

Doing a delta computation based on Dispatcher::approximateMonotonicTime effectively mixes time sources since timer registration uses a different monotonic timer internally (event_base_gettimeofday_cached).  Mixing time sources adds inaccuracies.  The result is not obviously better than max_configured-min_configured.